extensions/vi-mode/commands.lisp

```diff
@@ -275,21 +275,21 @@
 
 (define-operator vi-delete (start end type) ("<R>")
     (:move-point nil)
-  (let ((pos (point-charpos (current-point))))
+  (let ((pos (point-charpos (current-point)))
+        (ends-with-newline (char= (character-at end -1) #\Newline)))
     (if (visual-p)
         (visual-kill)
         (with-killring-context (:options (when (eq type :line) :vi-line))
           (kill-region-without-appending start end)))
     (when (and (eq type :line)
-               (eq 'vi-delete (command-name (this-command))))
-      (if (last-line-p (current-point))
-          (unless (first-line-p (current-point))
-            (delete-previous-char))
-          (delete-next-char))
-      (setf (point-charpos (current-point))
-            (max 0
-                 (min (1- (length (line-string (current-point)))) pos))))
+               (not ends-with-newline)
+               (not (= (position-at-point start) 1)))
+      (delete-previous-char))
     (when (eq 'vi-delete (command-name (this-command)))
+      (when (eq type :line)
+        (setf (point-charpos (current-point))
+              (max 0
+                   (min (1- (length (line-string (current-point)))) pos))))
```

Comment:
Direct change of point-charpos is deprecated because it breaks encapsulation.
How about `(line-offset (current-point) 0 charpos)`